File: html_minifier.go
// html_minifier.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"strings"
)

type HTMLMinifier struct {
	RemoveComments    bool
	PreserveWhitespace bool
	RemoveQuotes      bool
}

func NewHTMLMinifier(removeComments, preserveWhitespace, removeQuotes bool) *HTMLMinifier {
	return &HTMLMinifier{
		RemoveComments:    removeComments,
		PreserveWhitespace: preserveWhitespace,
		RemoveQuotes:      removeQuotes,
	}
}

func (m *HTMLMinifier) Minify(html string) string {
	// Remove comments
	if m.RemoveComments {
		re := regexp.MustCompile(`<!--.*?-->`)
		html = re.ReplaceAllString(html, "")
	}

	// Strip lines and collapse spaces
	lines := strings.Split(html, "\n")
	var cleaned []string
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed != "" || strings.Contains(line, "<") {
			cleaned = append(cleaned, trimmed)
		}
	}
	html = strings.Join(cleaned, "")

	if !m.PreserveWhitespace {
		// Remove spaces between tags
		re := regexp.MustCompile(`>\s+<`)
		html = re.ReplaceAllString(html, "><")
		// Collapse multiple spaces
		re = regexp.MustCompile(` {2,}`)
		html = re.ReplaceAllString(html, " ")
	}

	if m.RemoveQuotes {
		re := regexp.MustCompile(`=\s*"([^"]*)"`)
		html = re.ReplaceAllString(html, `=$1`)
		re = regexp.MustCompile(`=\s*'([^']*)'`)
		html = re.ReplaceAllString(html, `=$1`)
	}

	// Remove multiple newlines
	re := regexp.MustCompile(`\n\s*\n`)
	html = re.ReplaceAllString(html, "\n")
	return strings.TrimSpace(html)
}

func main() {
	var inputFile, outputFile string
	var noComments, preserveWhitespace, removeQuotes bool
	flag.StringVar(&inputFile, "i", "", "Input HTML file (or stdin)")
	flag.StringVar(&outputFile, "o", "", "Output file (or stdout)")
	flag.BoolVar(&noComments, "no-comments", false, "Keep comments")
	flag.BoolVar(&preserveWhitespace, "preserve-whitespace", false, "Preserve whitespace inside text")
	flag.BoolVar(&removeQuotes, "remove-quotes", false, "Remove quotes from simple attributes")
	flag.Parse()

	var content string
	if inputFile != "" {
		data, err := ioutil.ReadFile(inputFile)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
		content = string(data)
	} else {
		stat, _ := os.Stdin.Stat()
		if (stat.Mode() & os.ModeCharDevice) == 0 {
			scanner := bufio.NewScanner(os.Stdin)
			var sb strings.Builder
			for scanner.Scan() {
				sb.WriteString(scanner.Text())
				sb.WriteString("\n")
			}
			content = sb.String()
		} else {
			fmt.Fprintln(os.Stderr, "No input provided.")
			os.Exit(1)
		}
	}

	minifier := NewHTMLMinifier(!noComments, preserveWhitespace, removeQuotes)
	minified := minifier.Minify(content)

	if outputFile != "" {
		err := ioutil.WriteFile(outputFile, []byte(minified), 0644)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
	} else {
		fmt.Println(minified)
	}
}

File: html_minifier.js
// html_minifier.js
const fs = require('fs');
const readline = require('readline');

class HTMLMinifier {
    constructor(options = {}) {
        this.removeComments = options.removeComments !== false;
        this.preserveWhitespace = options.preserveWhitespace || false;
        this.removeQuotes = options.removeQuotes || false;
    }

    minify(html) {
        // Remove comments
        if (this.removeComments) {
            html = html.replace(/<!--[\s\S]*?-->/g, '');
        }

        // Remove leading/trailing whitespace from lines and join
        const lines = html.split('\n').map(line => line.trim()).filter(line => line !== '');
        html = lines.join('');

        if (!this.preserveWhitespace) {
            // Remove spaces between tags
            html = html.replace(/>\s+</g, '><');
            // Collapse multiple spaces
            html = html.replace(/ {2,}/g, ' ');
        }

        if (this.removeQuotes) {
            html = html.replace(/=\s*"([^"]*)"/g, '=$1');
            html = html.replace(/=\s*'([^']*)'/g, '=$1');
        }

        // Remove extra newlines
        html = html.replace(/\n\s*\n/g, '\n');
        return html.trim();
    }
}

function readStdin() {
    return new Promise((resolve) => {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            terminal: false
        });
        let data = '';
        rl.on('line', (line) => { data += line + '\n'; });
        rl.on('close', () => { resolve(data); });
    });
}

async function main() {
    const args = process.argv.slice(2);
    let inputFile = null;
    let outputFile = null;
    let noComments = false;
    let preserveWhitespace = false;
    let removeQuotes = false;

    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--no-comments') noComments = true;
        else if (args[i] === '--preserve-whitespace') preserveWhitespace = true;
        else if (args[i] === '--remove-quotes') removeQuotes = true;
        else if (args[i] === '-o') { outputFile = args[++i]; }
        else if (!inputFile) inputFile = args[i];
    }

    let content;
    if (inputFile) {
        content = fs.readFileSync(inputFile, 'utf8');
    } else {
        content = await readStdin();
    }

    const minifier = new HTMLMinifier({
        removeComments: !noComments,
        preserveWhitespace,
        removeQuotes
    });
    const minified = minifier.minify(content);

    if (outputFile) {
        fs.writeFileSync(outputFile, minified, 'utf8');
    } else {
        console.log(minified);
    }
}

main().catch(console.error);
